### PR TITLE
Allow access to all antecedent scores

### PR DESCRIPTION
**Pull request by gabbard**

### CODE REVIEW

File: allennlp/models/coreference_resolution/coref.py [hunk 1]
```diff
@@ -252,6 +264,9 @@ def forward(self,  # type: ignore
         output_dict = {"top_spans": top_spans,
                        "antecedent_indices": valid_antecedent_indices,
                        "predicted_antecedents": predicted_antecedents}
+        if keep_antecedent_alternatives and keep_antecedent_alternatives[0]:
```
Comment (berquist) [inline allennlp/models/coreference_resolution/coref.py]: Why is `if keep_antecedent_alternatives` not enough? The implicit `bool` check should handle `None` and having members simultaneously:
```python
In [284]: l1 = None

In [285]: l2 = []

In [286]: l3 = [2]

In [287]: if l1:
     ...:     print("apple")
     ...: elif l2:
     ...:     print("banana")
     ...: elif l3:
     ...:     print("pear")
     ...:
pear
```

File: allennlp/models/coreference_resolution/coref.py [hunk 2]
```diff
@@ -141,6 +146,13 @@ def forward(self,  # type: ignore
             A tensor of shape ``(batch_size, num_spans_to_keep)`` representing, for each top span, the
             index (with respect to antecedent_indices) of the most likely antecedent. -1 means there
             was no predicted link.
+        antecedent_scores : ``torch.FloatTensor``, optional
+            A tensor of shape ``(batch_size, num_spans_to_keep, 1 + max_antecedents)`` giving
+            the antecedent scores for each mention.  Each i-th batch element is associated with a
+            matrix whose the j-th row contains the antecedent scores for the j-th mention of that
+            batch (corresponding to top_spans).  The column k-th column contains the score for
```
Comment (berquist) [inline allennlp/models/coreference_resolution/coref.py]:
```suggestion
            batch (corresponding to top_spans), and k-th column contains the score for
```

File: allennlp/models/coreference_resolution/coref.py (hunk 1 above)
Comment (gabbard) [reply allennlp/models/coreference_resolution/coref.py]: But `bool([False])` is `True`, which is the situation here.  (one the initial instance, you have a field `keep_antecedent_alternatives` which maps to a boolean value, but the vectorization code in AllenNLP wraps the boolean value in a list by the time it gets here

Comment (berquist) [reply allennlp/models/coreference_resolution/coref.py]: Ah, I missed "containing the single value ``True``".

Comment (gabbard) [reply allennlp/models/coreference_resolution/coref.py]: But that does remind me that this checks only if the first thing in a batch has requested coref_scores, and we want to return them if *any* do. Just forced-pushed an amended commit with a fix.

File: allennlp/models/coreference_resolution/coref.py [hunk 3]
```diff
@@ -141,6 +146,13 @@ def forward(self,  # type: ignore
...
+            the (k-1)-th mention being the antecedent of the j-th mention. The first column
+            (index 0) contains the score for the j-th mention having no antecedent.
```
Comment (berquist) [inline allennlp/models/coreference_resolution/coref.py]: To make sure we are talking about the matrix that results from indexing into the tensor:
```suggestion
            (index k = 0) contains the score for the j-th mention having no antecedent.
```

File: allennlp/models/coreference_resolution/coref.py [hunk 4]
```diff
@@ -127,6 +129,9 @@ def forward(self,  # type: ignore
         span_labels : ``torch.IntTensor``, optional (default = None)
             A tensor of shape (batch_size, num_spans), representing the cluster ids
             of each span, or -1 for those which do not appear in any clusters.
+        output_alternative_antecedents: ``Optional[ScatterableList]`` - if non-`None` and
+            containing the single value ``True``, the output dictionary will contain antecedent
```
Comment (berquist) [inline allennlp/models/coreference_resolution/coref.py]: This should now mention that `output_alternative_antecedents` may have length greater than one.

**Review by berquist — CHANGES_REQUESTED** [review]

File: allennlp/models/coreference_resolution/coref.py (hunk 4 above)
Comment (gabbard) [reply allennlp/models/coreference_resolution/coref.py]: Fixed

**Review by berquist — APPROVED** [review]

**Pull request merged**